File: test1/kalkulator.py
import math
import cmath
import numpy as np
from statistics import mean, median, mode
import tkinter as tk
from tkinter import messagebox
import json
from datetime import datetime
import matplotlib.pyplot as plt
import sympy as sp

class Calculator:
    def __init__(self):
        self.history = []
        self.default_units = 'metric'  # User preference for units

    def add(self, x, y):
        result = x + y
        self.history.append(f"{datetime.now()}: {x} + {y} = {result}")
        return result

    def subtract(self, x, y):
        result = x - y
        self.history.append(f"{datetime.now()}: {x} - {y} = {result}")
        return result

    def multiply(self, x, y):
        result = x * y
        self.history.append(f"{datetime.now()}: {x} * {y} = {result}")
        return result

    def divide(self, x, y):
        if y == 0:
            return "Error! Division by zero."
        result = x / y
        self.history.append(f"{datetime.now()}: {x} / {y} = {result}")
        return result

    def square_root(self, x):
        result = math.sqrt(x)
        self.history.append(f"{datetime.now()}: √{x} = {result}")
        return result

    def power(self, x, y):
        result = x ** y
        self.history.append(f"{datetime.now()}: {x} ^ {y} = {result}")
        return result

    def percentage(self, x, y):
        result = (x * y) / 100
        self.history.append(f"{datetime.now()}: {y}% of {x} = {result}")
        return result

    def sin(self, x):
        result = math.sin(math.radians(x))
        self.history.append(f"{datetime.now()}: sin({x}) = {result}")
        return result

    def cos(self, x):
        result = math.cos(math.radians(x))
        self.history.append(f"{datetime.now()}: cos({x}) = {result}")
        return result

    def tan(self, x):
        result = math.tan(math.radians(x))
        self.history.append(f"{datetime.now()}: tan({x}) = {result}")
        return result

    def log(self, x, base=math.e):
        if x <= 0:
            return "Error! Logarithm of non-positive number."
        result = math.log(x, base)
        self.history.append(f"{datetime.now()}: log({x}, {base}) = {result}")
        return result

    def matrix_add(self, A, B):
        return np.add(A, B)

    def matrix_subtract(self, A, B):
        return np.subtract(A, B)

    def matrix_multiply(self, A, B):
        return np.matmul(A, B)

    def calculate_expression(self, expression):
        try:
            result = eval(expression)
            self.history.append(f"{datetime.now()}: {expression} = {result}")
            return result
        except Exception as e:
            return f"Error: {str(e)}"

    def clear_history(self):
        self.history.clear()
        return "History cleared."

    def get_history(self):
        return self.history

    def calculate_statistics(self, data):
        return {
            'mean': mean(data),
            'median': median(data),
            'mode': mode(data)
        }

    def convert_units(self, value, from_unit, to_unit):
        conversions = {
            'meters_to_feet': value * 3.28084,
            'feet_to_meters': value / 3.28084,
            'celsius_to_fahrenheit': (value * 9/5) + 32,
            'fahrenheit_to_celsius': (value - 32) * 5/9,
            'kilograms_to_pounds': value * 2.20462,
            'pounds_to_kilograms': value / 2.20462,
        }
        return conversions.get(f"{from_unit}_to_{to_unit}", "Invalid conversion")

    def save_history(self, filename):
        with open(filename, 'w') as f:
            json.dump(self.history, f)

    def load_history(self, filename):
        with open(filename, 'r') as f:
            self.history = json.load(f)

    def plot_function(self, func, x_range):
        x = np.linspace(x_range[0], x_range[1], 100)
        y = eval(func)
        plt.plot(x, y)
        plt.title(f"Plot of {func}")
        plt.xlabel("x")
        plt.ylabel("f(x)")
        plt.grid()
        plt.show()

    def symbolic_calculation(self, expression):
        x = sp.symbols('x')
        expr = sp.sympify(expression)
        return expr

def main():
    calc = Calculator()
    print("Advanced Calculator")
    print("Type 'exit' to quit")
    print("Type 'history' to view calculation history")
    print("Type 'clear' to clear history")
    print("Type 'save' to save history to a file")
    print("Type 'load' to load history from a file")
    print("Type 'plot' to plot a function")
    print("Type 'symbolic' to perform symbolic calculations")

    while True:
        expression = input("Enter expression: ")

        if expression.lower() == 'exit':
            break
        elif expression.lower() == 'history':
            print("Calculation History:")
            for entry in calc.get_history():
                print(entry)
            continue
        elif expression.lower() == 'clear':
            print(calc.clear_history())
            continue
        elif expression.lower() == 'save':
            filename = input("Enter filename to save history: ")
            calc.save_history(filename)
            print(f"History saved to {filename}.")
            continue
        elif expression.lower() == 'load':
            filename = input("Enter filename to load history: ")
            calc.load_history(filename)
            print(f"History loaded from {filename}.")
            continue
        elif expression.lower() == 'plot':
            func = input("Enter function to plot (e.g., np.sin(x)): ")
            x_range_input = input("Enter x range (e.g., -10, 10): ")
            try:
                x_range = [float(x) for x in x_range_input.split(",")]
                if len(x_range) != 2:
                    raise ValueError("Please provide exactly two values.")
            except ValueError as e:
                print(f"Invalid input for x range: {e}")
                continue
            calc.plot_function(func, x_range)
            continue
        elif expression.lower() == 'symbolic':
            sym_expr = input("Enter symbolic expression (e.g., x**2 + 2*x + 1): ")
            result = calc.symbolic_calculation(sym_expr)
            print(f"Symbolic result: {result}")
            continue

        result = calc.calculate_expression(expression)
        print(f"Result: {result}")

if __name__ == "__main__":
    main()

File: test3/Kalkulator.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Kalkulator {
    private List<String> history;

    public Kalkulator() {
        history = new ArrayList<>();
    }

    public double add(double x, double y) {
        double result = x + y;
        history.add(x + " + " + y + " = " + result);
        return result;
    }

    public double subtract(double x, double y) {
        double result = x - y;
        history.add(x + " - " + y + " = " + result);
        return result;
    }

    public double multiply(double x, double y) {
        double result = x * y;
        history.add(x + " * " + y + " = " + result);
        return result;
    }

    public double divide(double x, double y) {
        if (y == 0) {
            System.out.println("Error! Division by zero.");
            return Double.NaN;
        }
        double result = x / y;
        history.add(x + " / " + y + " = " + result);
        return result;
    }

    public double squareRoot(double x) {
        double result = Math.sqrt(x);
        history.add("√" + x + " = " + result);
        return result;
    }

    public double power(double x, double y) {
        double result = Math.pow(x, y);
        history.add(x + " ^ " + y + " = " + result);
        return result;
    }

    public double sin(double x) {
        double result = Math.sin(Math.toRadians(x));
        history.add("sin(" + x + ") = " + result);
        return result;
    }

    public double cos(double x) {
        double result = Math.cos(Math.toRadians(x));
        history.add("cos(" + x + ") = " + result);
        return result;
    }

    public double tan(double x) {
        double result = Math.tan(Math.toRadians(x));
        history.add("tan(" + x + ") = " + result);
        return result;
    }

    public double log(double x) {
        if (x <= 0) {
            System.out.println("Error! Logarithm of non-positive number.");
            return Double.NaN;
        }
        double result = Math.log(x);
        history.add("log(" + x + ") = " + result);
        return result;
    }

    public void printHistory() {
        System.out.println("Calculation History:");
        for (String entry : history) {
            System.out.println(entry);
        }
    }

    public static void main(String[] args) {
        Kalkulator calc = new Kalkulator();
        Scanner scanner = new Scanner(System.in);
        String command;

        System.out.println("Advanced Calculator");
        System.out.println("Type 'exit' to quit");
        System.out.println("Type 'history' to view calculation history");
        System.out.println("Available operations: +, -, *, /, sqrt, ^, sin, cos, tan, log");

        while (true) {
            System.out.print("Enter command: ");
            command = scanner.nextLine();

            if (command.equalsIgnoreCase("exit")) {
                break;
            } else if (command.equalsIgnoreCase("history")) {
                calc.printHistory();
                continue;
            }

            String[] parts = command.split(" ");
            if (parts.length < 3) {
                System.out.println("Invalid command. Please enter a valid operation.");
                continue;
            }

            double x = Double.parseDouble(parts[0]);
            String operator = parts[1];
            double y = Double.parseDouble(parts[2]);
            double result = 0;

            switch (operator) {
                case "+":
                    result = calc.add(x, y);
                    break;
                case "-":
                    result = calc.subtract(x, y);
                    break;
                case "*":
                    result = calc.multiply(x, y);
                    break;
                case "/":
                    result = calc.divide(x, y);
                    break;
                case "sqrt":
                    result = calc.squareRoot(x);
                    break;
                case "^":
                    result = calc.power(x, y);
                    break;
                case "sin":
                    result = calc.sin(x);
                    break;
                case "cos":
                    result = calc.cos(x);
                    break;
                case "tan":
                    result = calc.tan(x);
                    break;
                case "log":
                    result = calc.log(x);
                    break;
                default:
                    System.out.println("Invalid operator.");
                    continue;
            }
            System.out.println("Result: " + result);
        }
        scanner.close();
    }
}
